### ISSUE TITLE
Final rewards from model evaluation are different than final rewards from model callbacks (proper format)

### ISSUE THREAD

**Issue by george-adams1** [DLR-RM/stable-baselines3#1552]
### 🐛 Bug

Reference to old issue: https://github.com/DLR-RM/stable-baselines3/issues/1551

In the training env, I've set up callbacks every 10k timesteps where a snapshot of the model in its current state is tested on all of the in-sample training data. Several different metrics are calculated, most importantly reward. At the end of the training, during the last callback when the model is finished training, it gives a certain reward from the callback. However, when the model is finished training, and I take that exact same model and run it manually through an evaluation function through the exact same environment with the same data, the final reward is dramatically different from the reward for the same model at the same number of training timesteps. For example, a callback reward of 4e-4 and an evaluation reward of -4e-4. In both cases, the model is deterministic. 

Any idea why this may be the case?

### Code example

I think it will be pretty difficult to completely reproduce this issue since the full environment is quite large. The most help would be to look at the plots below and comment on why that may be the case. 

```python
import gymnasium as gym
import numpy as np
from gymnasium import spaces

from stable_baselines3 import PPO
from stable_baselines3.common.env_checker import check_env


class CustomEnv(gym.Env):

    def __init__(self):
        super().__init__()
        self.observation_space = spaces.Box(low=-np.inf, high=np.inf, shape=(obs_shape,))
        self.action_space = spaces.MultiDiscrete([2] * self.num_currencies) 
        self.action_space = spaces.MultiDiscrete([2] * self.num_currencies)

    def reset(self, seed=None, options=None):
        self.current_step = 0
        self.reward = 0
        self.total_asset = 0

         # records
        self.asset_memory = [self.initial_balance]
        self.cash_memory = [self.initial_balance]
        self.actions_memory = []
        self.date_memory = [self._get_date()]
        self.total_reward = 0

        return self._get_observation()

    def step(self, action):
        obs = self.observation_space.sample()

      # model makes action       

        self.actions_memory.append(action)

        self.asset_memory.append(self.total_asset)
        self.cash_memory.append(self.account_balance)
        self.date_memory.append(self._get_date())
        reward = self.calculate_reward()
        self.total_reward += reward

        return self._get_observation(), reward, self.done, {}

env = CustomEnv()
check_env(env)

model = PPO("MlpPolicy", env, verbose=1).learn(1000)
```

The evaluation function described above is as follows: 

```
    def evaluation(self, model, deterministic=True, IS_data=False):
        if IS_data:
            env = self.train_env

        total_rewards = 0
        test_obs = env.reset()
        for i in range(len(env.master_data)):
            action, _states = model.predict(test_obs, deterministic=deterministic)
            test_obs, rewards, done, info = env.step(action)
            total_rewards += rewards
        return total_rewards
```

Callback mean reward plot (timesteps vs reward): 
![photo_2023-06-13_14-16-54](https://github.com/DLR-RM/stable-baselines3/assets/51809360/a92e1b46-9b01-413c-b98d-99d67e133ed8)

Evaluation function mean reward plot (timesteps vs reward): 
![photo_2023-06-13_14-17-01](https://github.com/DLR-RM/stable-baselines3/assets/51809360/c744394c-0cb8-46f7-a0f8-9246f1a67c69)

### Relevant log output / Error message

```shell
There is nothing very relevant in the logs and no errors. Look at the two graphs above to get a better idea of the errors.
```

### System Info

- OS: Windows-10-10.0.22000-SP0 10.0.22000
- Python: 3.9.16
- Stable-Baselines3: 1.8.0
- PyTorch: 2.0.0
- GPU Enabled: False
- Numpy: 1.22.3
- Gym: 0.21.0
- Installed with pip

({'OS': 'Windows-10-10.0.22000-SP0 10.0.22000', 'Python': '3.9.16', 'Stable-Baselines3': '1.8.0', 'PyTorch': '2.0.0', 'GPU Enabled': 'False', 'Numpy': '1.22.3', 'Gym': '0.21.0'}, '- OS: Windows-10-10.0.22000-SP0 10.0.22000\n- Python: 3.9.16\n- Stable-Baselines3: 1.8.0\n- PyTorch: 2.0.0\n- GPU Enabled: False\n- Numpy: 1.22.3\n- Gym: 0.21.0\n')

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have provided a minimal working example to reproduce the bug
- [X] I have checked my env using the env checker
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by nollied** [comment]
@george-adams1 any response? did this make it into the 2.0.0 release?

**Comment by nollied** [comment]
cus i'm still experiencing the same issue.
